docs: add D2xx, FT8U2xxA, FT2xxB, FT2232 docs. (#3)
wanda-phi/prjspeedbump: docs/src/d2xx.md

# D2xx protocol

> D2xx is the name of protocol used by most FTDI USB devices.  It was originally made for the UART function of the FT8U100AX, and remains quite UART-oriented.  However, FTDI has reused it for all kinds of stream-based interfaces.

Reviewer: What is the meaning of the X in `FT8U100AX`? In `devices.md` you refer to this as `FT8U100A`, without the `X`.

Author: the last letter of an FTDI name is (usually) a package designator, 100-pin PQFP in this case (doesn't really matter as the FT8U100A doesn't come in any other package)

I'll remove the X for consistency

As you can see, FTDI used to reuse its VID:PID combinations for groups of similar devices for some time, but then decided to just start assigning new IDs for every new device.

Once something is known to be a FTD2xx device, further identification can be obtained from the `bcdDevice` field of USB descriptor.  The `bcdDevice` values are likewise listed on the [devices page](devices.md#d2xx-devices).  It is not clear whether the `bcdDevice` field may be overriden via EEPROM.

Even putting VID:PID and `bcdDevice` together, it is still not always possible to uniquely identify the device.  Usually, one can just live with the uncertainty.  However, there are some more checks that can be done for further identification:

- FT8U2xxA and FT2xxB can be told apart by attempting the `GET_LATENCY_TIMER` request — a failure implies A-series device, while a success implies B-series device

> - FT232R and FT245R can be told apart by an EEPROM field

Reviewer: ```suggestion
- FT232R and FT245R can be told apart by a field in the internal EEPROM, which the vendor claims to not be modifiable by the user
```

> - likewise, FT-X series devices can be told apart to some extend by another EEPROM field

Reviewer: ```suggestion
- likewise, FT-X series devices can be told apart to some extent by another EEPROM field
```

## Stream protocol

> Note: FT2xxR devices have configurable `wMaxPacketSize` field in the device descriptor.  There are devices in the wild where people have programmed the (invalid) value of 0 in that field.  In case a device is encountered with such value, it should be overriden in the driver to 64 bytes.

Reviewer: `wMaxPacketSize` is a field in the endpoint descriptor, so you need to be more specific. Is this about the control endpoint?

Author: bulk endpoint; fixed

### IN endpoint

> The IN endpoint of a channel is used to transfer byte stream data from the device to the user.  It involves a framing protocol: every packet transferred on the endpoint includes a header with modem and line status.  Packets are always at least 2 bytes long.  The packet format is as follows:

Reviewer: Note that the IN request can request a specific packet size. You should mention how (at least on the FT2232H, where I tested this) this behaves a bit differently than you might expect – if your requested packet size is lower than the maximum, and the on-device buffer contains more data that can fit in your requested size, then the device will respond with an error instead of sending a smaller packet.

Author: you actually cannot (which is why you should always make your buffer a multiple of max packet size); there is no "requested bytes" field in the `IN` token packet, so the device has no way of known your buffer size and can always reply with the max packet size it supports. if this happens to overflow your buffer, some part of the host software stack will materialize an error.

Reviewer: I got bamboozled by the tower of abstractions! >.<

- byte 0: modem status (shows the most recent state of the modem status input pins)
  - bit 0: always 1?
  - bit 1: always 0?
  - bit 2: always 0?
  - bit 3: always 0?

>   - TODO: allegedly high-speed devices have 0010 in low bits?

Reviewer: I think this also signals the expected max packet size of the IN endpoint?

Author: that'd be a very strange layering violation (and redundant with the endpoint descriptors)... but then, who knows

I'm punting this to hw testing, unless you happen to have some evidence from the driver?

  - bit 2: parity error
  - bit 3: framing error
  - bit 4: break interrupt
  - bit 5: TX holding register (???)
  - bit 6: TX (OUT endpoint) FIFO is empty
  - bit 7: FIFO error (???)
- bytes 2 and up (if any): actual stream data

As can be seen, the protocol is very UART-oriented.  The modem and line status bytes are present even for D2xx devices that have nothing to do with UART, though they are mostly dummied out in that case.  The TX FIFO empty bit is, however, generally useful.

TODO: kernel driver alleges that parity/framing errors are always associated with the last byte in the packet. Check this.

To avoid excessive load on the host, the device doesn't always transmit packets on the IN endpoint as soon as data is available.  Instead, the device will send a packet in the following circumstances:

- a full-sized packet has been filled
- a special "event" character has been received (if enabled by driver)
- data is available and the latency timer has expired
  - on FT8U2xxA (and FT8U100A?), the latency timer is always 16ms
  - on newer devices, the default latency timer is 16ms, but it can be configured to any value between 2ms and 255ms

> - TODO: line errors / modem status changes too? kernel alleges device will send an update packet every 40ms even with no activity

Reviewer: Wait, but the device doesn't just transmit packets, it needs to do so in response to an IN request from the host, right? Does it just respond with the USB equivalent of `-EAGAIN` in that case, to... uh, save bandwidth? Not sure if this makes any sense...

Author: yeah it gives you a NAK to lower interrupt load on the host, I think

### OUT endpoint

There are two variants of the protocol with respect to OUT endpoint handling.

The original variant is present on the FT8U100A only.  In this variant, every OUT endpoint packet has to be prefixed with a header.  The packet then has the following format:

- byte 0: header
  - bit 0: must be 1
  - bit 1: must be 0
  - bits 2-7: length of payload (ie. length of packet, not including the header byte)
- bytes 1 and up: actual stream data

Since the header is completely useless, it has been done away with in later devices.  In every device other than FT8U100A, the OUT endpoint is completely headerless and packets contain just the raw stream data.

## Control requests

The protocol includes the following control requests:

| request                    | `bmRequestType` | `bRequest` | `wValue`                  | `wIndex`                    | data             |
| -------------------------- | --------------- | ---------- | ------------------------- | --------------------------- | ---------------- |

> | `RESET`                    | `0x40`          | `0x00`     | reset kind                | channel                     | -                |

Reviewer: The driver I looked at does `RESET` with `wIndex` always set to 0.

Author: I think this is a bug; in any case, added as a TODO

| `READ_EEPROM`              | `0xc0`          | `0x90`     | 0                         | word address                | 2 bytes          |
| `WRITE_EEPROM`             | `0x40`          | `0x91`     | word value                | word address                | -                |
| `ERASE_EEPROM`             | `0x40`          | `0x92`     | 0                         | 0                           | -                |

Most control requests operate on a channel.  The channel is specified as follows:

- for single-channel devices: channel is 0
- for multi-channel devices: the channel indices start at 1 (ie. channel number in control request is USB interface index + 1)
  - TODO: allegedly, specifying 0 means "all channels"?

The channel is usually stuffed directly into `wIndex`.  However, there are variations to keep you on your toes.

### `RESET`

Resets the device or purges IN/OUT endpoint buffers on device.

- applicable to: all D2xx devices
- `bmRequestType`: `0x40`
- `bRequest`: `0x00`
- `wValue`: kind of reset to perform
  - 0: reset the serial engine
    - purge both IN and OUT endpoints
    - disable event char and set it to `0x0d` (`'\r'`)

>     - diable flow control

Reviewer: ```suggestion
    - disable flow control
```

    - clear DTR and RTS
  - 1: purge OUT endpoint (nuke all buffered data)
  - 2: purge IN endpoint
- `wIndex`: channel

TODO: determine *exactly* what the reset actually resets; the above list is according to the kernel sources.

### `SET_MODEM_CTRL`

Changes the state of DTR and RTS lines.

- applicable to: all D2xx devices (but may be a no-op in non-UART modes)
- `bmRequestType`: `0x40`
- `bRequest`: `0x01`
- `wValue`:
  - bit 0: change DTR (if not set, DTR will be unchanged)
  - bit 1: change RTS
  - bit 8: new DTR value

>   - bit 9: new RTS value

Reviewer: AFAIK this is only correct if you view the field as big endian and nobody does that. The upper bits are "should this change", the lower bits are the new value.

Author: okay that's a plain fuckup on my side

- `wIndex`: channel

### `SET_FLOW_CTRL`

Changes hardware flow control mode.

- applicable to: all D2xx devices (but may be a no-op in non-UART modes)
- `bmRequestType`: `0x40`
- `bRequest`: `0x02`
- `wValue`:
  - bits 0-7: XON character
  - bits 8-15: XOFF character
- `wIndex`:
  - bits 0-7: channel
  - bit 8: enable RTS/CTS flow control
  - bit 9: enable DTR/DSR flow control

>   - bit 10: enable XON/XOFF flow control

Reviewer: oeirgjsoidfjso TODO what happens if you set more than one bit at once

Author: I don't think we need a TODO for every possible protocol violation tbh

Reviewer: well this does read like you can enable more than one. I can imagine it responding to both RTS/CTS and XON/XOFF, for example.

    - bits 0-7: channel
    - bits 8-9: bits 16-17 of the divisor

The format of the divisor varies depending on the exact device.

#### FT8U100A

For FT8U100A, the divisor is an enum, selecting the baud rate from the following fixed list:

- 0: 300 baud
- 1: 600
- 2: 1200
- 3: 2400
- 4: 4800
- 5: 9600
- 6: 19200
- 7: 38400
- 8: 57600
- 9: 115200

#### FT8U232A, FT8U245A

These devices run on 48MHz base clock.  The base clock is divided by a fractional divisor to obtain the channel clock.  The channel clock is then further divided by 16 to obtain the baud rate.

The divisor is encoded as follows:

- bits 0-13: integer part (must be at least 2, except for the special case below)
- bits 14-15: fractional part, encoded
  - 0: .0
  - 1: .5
  - 2: .25
  - 3: .125

There is also a special divisor with dedicated encoding:

- `0x0000` encodes a special divisor of 1 (corresponding to 3000000 baud rate)

#### FT232B, FT245B, FT232R, FT245R, FT2232[CDL], FT-X series

These devices are similar to FT8U2xxA, with minor extensions.  The encoding scheme is backwards-compatible.

The divisor is encoded as follows:

- bits 0-13: integer part (must be at least 2, except for the special cases below)

> - bits 14-16: fractional part, encoded

Reviewer: I'd mention the whole "upper bits stuffed in `wIndex` because the divisor is 17 bits wide" thing once again here, because at first glance this reads like an off-by-one.

- applicable to: everything except FT8U100A, FT8U232A, FT8U245A
- `bmRequestType`: `0xc0`
- `bRequest`: `0x0a`
- `wValue`: 0
- `wIndex`: channel
- `wLength`: 1
- data: latency timer (1 byte)

### `SET_BITMODE`

Sets a special mode on the channel, overriding whatever base mode it has.

- applicable to: FT232B, FT245B, FT232R, FT245R, FT2232[CDL], FT2232H, FT4232H, FT232H, FT-X series
  - async bitbang: all of the above devices
  - MPSSE: FT2232[CDL] (channel A only), FT2232H, FT4232H (channel A and B only), FT232H
  - sync bitbang: FT232R, FT245R, FT2232[CDL], FT2232H, FT4232H, FT232H, FT-X series
  - MCU host bus: FT2232[CDL], FT2232H
  - fast opto-isolated serial: FT2232[CDL], FT2232H, FT232H
  - CBUS bit-bang: FT232R, FT232H, FT-X series
  - synchronous 245-style FIFO: FT2232H, FT232H
  - FT1248: FT232H maybe?!?
- `bmRequestType`: `0x40`
- `bRequest`: `0x09`
- `wValue`:
  - bits 0-7: depends on mode
    - async and sync bit-bang: mask of which data pins are outputs (0 is input, 1 is output)
    - CBUS bit-bang:
      - bits 0-3: output data
      - bits 4-7: output pin mask (0 is input, 1 is output)

>     - others: unused?

Reviewer: are you sure this is unused in MPSSE mode?

Author: absolutely not, hence the question mark

however, I'm going to note that this wouldn't make much sense as MPSSE allows you to manipulate 16 pins, while this field only has room for 8.

- `wIndex`: channel (or not, it's vendor-defined after all)
- `wLength`: up to `0x80`

If exactly 1 byte is sent to the device, it is put in the high byte of `wValue`, and `wLength` is set to 0.  If more than 1 byte is sent to the device, they are all put in the data payload, and the high byte of `wValue` is unused.

The request set appears to be specific to the exact device.  For the FT4222H request list, see [TODO].

### `READ_EEPROM`

Reads a 16-bit word from the EEPROM.

- applicable to: FT8U100A?, FT8U232A, FT8U245A, FT232B, FT245B, FT232R, FT245R, FT2232[CDL], FT2232H, FT4232H, FT232H, FT-X series
- `bmRequestType`: `0xc0`
- `bRequest`: `0x90`
- `wValue`: 0
- `wIndex`: word address to read; note that the address is counted in 16-bit words, not bytes
- `wLength`: 2
- data: word from EEPROM (little-endian)

The format of the EEPROM depends on the device.  See the documentation for the particular device.

### `WRITE_EEPROM`

Writes a 16-bit word into the EEPROM.

- applicable to: FT8U100A?, FT8U232A, FT8U245A, FT232B, FT245B, FT232R, FT245R, FT2232[CDL], FT2232H, FT4232H, FT232H, FT-X series
- `bmRequestType`: `0x40`
- `bRequest`: `0x91`
- `wValue`: word to write to the EEPROM

> - `wIndex`: word address to write; note that the address is counted in 16-bit words, not bytes

Reviewer: Should the even-odd address on 32-bit internal EEPROMs thing be documented here?

Author: I'll leave this to FT232R docs

Author: On second thought, I'm going to mention this here anyway, and write it down in detail in FT232R docs.

#### FT232B, FT245B, FT232R, FT245R, FT2232[CDL], FT-X series

These devices are similar to FT8U2xxA, with minor extensions.  The encoding scheme is backwards-compatible.

The divisor is encoded into a 17-bit number (with top bit overflowing into the `wIndex` field) as follows:

- bits 0-13: integer part (must be at least 2, except for the special cases below)
- bits 14-16: fractional part, encoded
  - 0: .0
  - 1: .5
  - 2: .25
  - 3: .125
  - 4: .375
  - 5: .675
  - 6: .75
  - 7: .875

There are also two special divisors with dedicated encoding:

- `0x0000` encodes a special divisor of 1 (corresponding to 3000000 baud rate)
- `0x0001` encodes a special divisor of 1.5 (corresponding to 2000000 baud rate)

In addition to providing the UART baud rate, the channel clock (not divided by 16) is also used for other modes such as bitbang and MPSSE.

#### FT2232H, FT4232H, FT232H (and all their variants)

These devices are USB high-speed peripherials, and support a higher baud rate and channel clock range.  The base clock of the device is 120MHz.  The divisor encoding is an extension of the one used in FT232B, and is backwards compatible.

> The divisor is encoded into an 18-bit number (with top bit overflowing into the `wIndex` field) as follows:

Reviewer: ```suggestion
The divisor is encoded into an 18-bit number (with top bits overflowing into the `wIndex` field) as follows:
```
